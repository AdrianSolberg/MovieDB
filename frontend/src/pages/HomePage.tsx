import { useEffect, useState } from "react";
import SortAndFilterPanel from "../components/SortAndFilterPanel";
import MovieList from "../components/MovieList";
import { MoviePoster } from "../types/movieTypes";
import SearchBar from "../components/SearchBar";
import { gql, useQuery, useReactiveVar } from "@apollo/client";
import { filtersVar } from "@/utils/cache";

const GET_MOVIES = gql`
    query GetMovies($skip: Int, $limit: Int, $filters: MovieFilters) {
        movies(skip: $skip, limit: $limit, filters: $filters) {
            _id
            title
            vote_average
            release_date
            runtime
            poster_path
        }
    }
`;

type MoviePosterRaw = Omit<MoviePoster, "release_date"> & {
    release_date: string;
};

interface GetMoviesData {
    movies: MoviePosterRaw[];
}

function HomePage() {
    const [movies, setMovies] = useState<MoviePoster[]>([]);
    const filters = useReactiveVar(filtersVar);

    const { data, loading, error, fetchMore } = useQuery<GetMoviesData>(
        GET_MOVIES,
        {
            variables: {
                skip: 0,
                limit: 20,
                filters: {
                    genre: filters.Genre || [],
                    rating: filters.Rating || [],
                    releaseYear: filters["Release Year"] || [],
                    language: filters.Language || [],
                    status: filters.Status || [],
                    runtime: filters.Runtime || [],
                },
            },
        }
    );

    useEffect(() => {
        if (data?.movies) {
            const moviesWithDate = data.movies.map((movie) => ({
                ...movie,
                release_date: new Date(movie.release_date),
            }));
            setMovies(moviesWithDate as MoviePoster[]);
        }
    }, [data]);

    if (loading) {
        return (
            <main className="mt-2 w-dvw text-center">
                <h1 className="text-2xl">Loading...</h1>
            </main>
        );
    }

    if (error) {
        return (
            <main className="mt-2 w-dvw text-center">
                <h1 className="text-2xl">Something went wrong!</h1>
                <h2 className="text-primary">Try to refresh</h2>
            </main>
        );
    }

    const loadMore = () => {
        fetchMore({
            variables: { skip: movies.length },
        });
    };

    return (
        <main>
            <section className="flex flex-row gap-2 p-4">
                <SortAndFilterPanel />
                <SearchBar />
            </section>
            <MovieList movies={movies} loadMore={loadMore} />
        </main>
    );
}

export default HomePage;
