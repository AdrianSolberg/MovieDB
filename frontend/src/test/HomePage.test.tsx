import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import HomePage from "../pages/HomePage";
import { vi } from "vitest";

vi.mock("../components/SortAndFilterPanel", () => ({
    default: vi.fn(() => <div data-testid="sort-and-filter-panel" />),
}));

vi.mock("../components/SearchBar", () => ({
    default: vi.fn(() => <div data-testid="search-bar" />),
}));

vi.mock("../components/MovieList", () => ({
    default: vi.fn(() => <div data-testid="movie-list" />),
}));

describe("HomePage Snapshots", () => {
    it("matches snapshot", () => {
        const { asFragment } = render(<HomePage />);
        expect(asFragment()).toMatchSnapshot();
    });
});

describe("HomePage", () => {
    it("renders SortAndFilterPanel, SearchBar, and MovieList", () => {
        render(<HomePage />);

        const sortAndFilterPanel = screen.getByTestId("sort-and-filter-panel");
        const searchBar = screen.getByTestId("search-bar");
        const movieList = screen.getByTestId("movie-list");

        expect(sortAndFilterPanel).toBeInTheDocument();
        expect(searchBar).toBeInTheDocument();
        expect(movieList).toBeInTheDocument();
    });
});
