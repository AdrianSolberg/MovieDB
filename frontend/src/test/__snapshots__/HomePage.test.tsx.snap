// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomePage > matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="flex flex-row gap-2 p-4"
  >
    <div
      data-testid="sort-and-filter-panel"
    />
    <div
      data-testid="search-bar"
    />
  </section>
  <div
    data-testid="movie-list"
  />
</DocumentFragment>
`;
